### ISSUE TITLE
node_children_visits should be within sqrt

### ISSUE THREAD

**Issue by schwgHao** [NohTow/PPL-MCTS#5]
When calculating `node_policy_score`  in line 422 of $mcts\\_rollout.py$, `node_children_visits` is outside `sqrt`, which is inconsistent with the formula in the original paper. Any consideration here? 
Thanks a lot.

**Comment by NohTow** [comment]
Hello,

In the paper, in the Section 3 where we describe the algorithm, in the selection step, we have: $P U C T(i)=\frac{s_i}{n_i}+c_{p u c t} * p_\theta\left(x_i \mid x_{1: t-1}\right) * \frac{\sqrt{N_i}}{1+n_i}$. Note that the `sqrt` is only applied to $N_i$.

On [line 422](https://github.com/NohTow/PPL-MCTS/blob/5f2b8ee96fbe14612d70dd7cb0d085d971dd476b/mcts_rollout.py#L422), we have `node_policy_score = np.sqrt(node_visits[:, None]) * self._pb_c_init * node_children_prior / (node_children_visits + 1)`, where `sqrt` is applied to `node_visits` ( $N_i$ ) and not to `node_children_visits` ( $n_i$ ), as described in the paper. This definition is also coherent with the definition of [Silver et al., 2017], so I am unsure where the confusion come from or which "original paper" you are referring to.

Let me know if I missed something.

**Comment by schwgHao** [comment]
> 
Hi,
Thanks for your quick reply.
I referred to subsection 3.2.3 in _Generative Cooperative Networks for Natural Language Generation_, another one of your excellent series papers. I thought they were the same but actually they are not. Can I know why the two have different forms when describing UCT. Thanks again.

**Comment by NohTow** [comment]
> I referred to subsection 3.2.3 in _Generative Cooperative Networks for Natural Language Generation_

Oh, I see, right, nice catch !
The code base used for GCN is not the same as this one and unfortunately, I do not have access to it so I cannot be sure about the answer, but here is all the information I gathered:

- In [SelfGAN](https://arxiv.org/pdf/2106.06363.pdf), the paper on which the code base of GCN is based, the definition is consistent with ours and Silver's
- The difference is (most likely) due to the fact that the one who write the definition within the paper is not the same who was coding
- Only applying the sqrt to $N_i$ is pretty common in different MCTS applications

To sum up: I think the code base of GCN use the same definition as PPL-MCTS and the difference in the paper is due to one of the author writing from the original UCT definition rather than the one we used. I am not sure if there is much of a difference in practice (maybe one of them is more theoretically backed), but in any case MCTS is working fine with this definition (see our studies, AlphaGo, ...). However, I cannot be 100% sure which one has been used in GCN (but I am *pretty* sure it was this one).

Hope it helps.

Edit: the one who wrote the formula in GCN also thinks this is most certainly a typo

**Comment by schwgHao** [comment]
> > I referred to subsection 3.2.3 in _Generative Cooperative Networks for Natural Language Generation_
> 
> Oh, I see, right, nice catch ! The code base used for GCN is not the same as this one and unfortunately, I do not have access to it so I cannot be sure about the answer, but here is all the information I gathered:
> 
> * In [SelfGAN](https://arxiv.org/pdf/2106.06363.pdf), the paper on which the code base of GCN is based, the definition is consistent with ours and Silver's
> * The difference is (most likely) due to the fact that the one who write the definition within the paper is not the same who was coding
> * Only applying the sqrt to Ni is pretty common in different MCTS applications
> 
> To sum up: I think the code base of GCN use the same definition as PPL-MCTS and the difference in the paper is due to one of the author writing from the original UCT definition rather than the one we used. I am not sure if there is much of a difference in practice (maybe one of them is more theoretically backed), but in any case MCTS is working fine with this definition (see our studies, AlphaGo, ...). However, I cannot be 100% sure which one has been used in GCN (but I am _pretty_ sure it was this one).
> 
> Hope it helps.
> 
> Edit: the one who wrote the formula in GCN also thinks this is most certainly a typo

Thanks for the clarification. It really helps a lot. Have a nice day!
